### ISSUE TITLE
float64 `log1p` on CPU is up to 129 ulp off near x ≈ -0.414 (also affects `atanh`)

### ISSUE THREAD

**Issue by timoklein** [jax-ml/jax#39707]
## Description

**NOTE**: The issue was identified by Claude. I double-checked and verified it on a different machine.

On the CPU backend, float64 `jnp.log1p` is inaccurate for arguments between roughly -0.53 and
-0.28. The error grows toward -0.4142 (where $1+x \approx 2-\sqrt{2}$ ) and peaks at 129 ulp. Outside the window, it is fine (below 1.5 ulp). NumPy's `log1p` is within 1 ulp on the same grid.

GPU is unaffected (≤ 2 ulp), float32 is unaffected (≤ 3 ulp), and jit vs. eager makes no difference.

**Discovery**: I ran into this when comparing the numerical stability of hyperbolic math operations for my own JAX-based library vs. PyTorch competitors on CPU. For example, the closed-form formula of the logarithmic map at the origin for the Poincaré Ball uses `jnp.atanh` (which uses `jnp.log1p`), and is therefore substantially less accurate than PyTorch and NumPy. float64 precision is often used in hyperbolic deep learning because the transcendental operations involved can otherwise lead to numerical breakdown.

## Reproduction

```python
import numpy as np, jax, jax.numpy as jnp
jax.config.update("jax_enable_x64", True)

y = np.float64(-0.4141892)
# Will produce different rounding than NumPy on CPU
print(jnp.log1p(y))
print(np.log1p(np.longdouble(y))) 

x = np.linspace(0.25, 0.60, 2_000_001)
ref = np.log1p(-x.astype(np.longdouble)).astype(np.float64)
err = np.abs(np.asarray(jnp.log1p(jnp.asarray(-x))) - ref) / np.spacing(np.abs(ref))
print(err.max(), (-x)[err.argmax()])           # 129.0 at -0.4141892
err_np = np.abs(np.log1p(-x) - ref) / np.spacing(np.abs(ref))
print(err_np.max())                            # 1.0
```

## What's affected

- `jnp.log1p`, `jnp.atanh`

### System info (python version, jaxlib version, accelerator, etc.)

Initial:
```
jax:    0.11.0
jaxlib: 0.11.0
numpy:  2.5.1
python: 3.13.8 (main, Oct  7 2025, 15:35:51) [Clang 20.1.4 ]
device info: NVIDIA A100-PCIE-40GB-2, 2 local devices"
process_count: 1
platform: uname_result(system='Linux', node='rey', release='6.8.0-54-generic', version='#56-Ubuntu SMP PREEMPT_DYNAMIC Sat Feb  8 00:37:57 UTC 2025', machine='x86_64')

$ nvidia-smi
Tue Aug  4 09:44:44 2026
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.216.03             Driver Version: 535.216.03   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-PCIE-40GB          Off | 00000000:5E:00.0 Off |                    0 |
| N/A   25C    P0              34W / 250W |    431MiB / 40960MiB |      4%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-PCIE-40GB          Off | 00000000:86:00.0 Off |                    0 |
| N/A   27C    P0              36W / 250W |    424MiB / 40960MiB |      3%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   1578054      C   ...rbolix-submission/.venv/bin/python3      422MiB |
|    1   N/A  N/A   1578054      C   ...rbolix-submission/.venv/bin/python3      416MiB |
+---------------------------------------------------------------------------------------+
```

Verified:
```
jax:    0.9.1
jaxlib: 0.9.1
numpy:  2.4.3
python: 3.12.7 (main, Oct 16 2024, 04:37:19) [Clang 18.1.8 ]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='cheesecake', release='6.8.0-136-generic', version='#136-Ubuntu SMP PREEMPT_DYNAMIC Wed Jul  1 21:53:05 UTC 2026', machine='x86_64')
```

**Comment by ifrahgul** [comment]
Hi! I am very interested in fixing this accuracy issue for float64 log1p on the CPU backend near x ≈ -0.414. Correcting this precision rounding error for log1p and atanh looks like a great numerical improvement for JAX. Since this issue is unassigned, could you please assign it to me?

**Comment by ayaka14732** [comment]
I have proposed a fix in https://github.com/openxla/xla/pull/46765

**Comment by timoklein** [comment]
That went really quickly! Thanks @ayaka14732 !

**Comment by ayaka14732** [comment]
@jakevdp Can you find someone to approve my PR?

**Comment by ayaka14732** [comment]
Should have been fixed at HEAD.
